icicle/tests/test_hash_api.cpp

```diff
@@ -61,34 +68,140 @@ class HashApiTest : public ::testing::Test
+
+  for (const auto& device : s_registered_devices) {
+    ICICLE_LOG_DEBUG << "Keccak256 test on device=" << device;
+    ICICLE_CHECK(icicle_set_device(device));
+    // Create Keccak-256 hash object and hash
+    auto keccak256 = Keccak256::create();
+    ICICLE_CHECK(keccak256.hash(input.data(), input.size() / config.batch, config, output.get()));
+    // Convert the output do a hex string and compare to expected output string
+    std::string output_as_str = voidPtrToHexString(output.get(), output_size);
+    ASSERT_EQ(output_as_str, expected_output_0);
+    output_as_str = voidPtrToHexString(output.get() + output_size, output_size);
+    ASSERT_EQ(output_as_str, expected_output_1);
+  }
+}
+
+TEST_F(HashApiTest, KeccakLarge)
+{
+  auto config = default_hash_config();
+  config.batch = 1 << 8;
+  const unsigned chunk_size = 1 << 13; // 8KB chunks
+  const unsigned total_size = chunk_size * config.batch;
+  auto input = std::make_unique<std::byte[]>(total_size);
+  randomize((uint64_t*)input.get(), total_size / sizeof(uint64_t));
+
+  const uint64_t output_size = 32;
+  auto output_main = std::make_unique<std::byte[]>(output_size * config.batch);
+  auto output_main_case_2 = std::make_unique<std::byte[]>(output_size * config.batch);
+  auto output_ref = std::make_unique<std::byte[]>(output_size * config.batch);
+
+  ICICLE_CHECK(icicle_set_device(s_reference_target));
+  auto keccakCPU = Keccak256::create();
+  START_TIMER(cpu_timer);
+  ICICLE_CHECK(keccakCPU.hash(input.get(), chunk_size, config, output_ref.get()));
+  END_TIMER(cpu_timer, "CPU Keccak large time", true);
+
+  ICICLE_CHECK(icicle_set_device(s_main_target));
+  auto keccakCUDA = Keccak256::create();
+
+  // test with host memory
+  START_TIMER(cuda_timer);
+  config.are_inputs_on_device = false;
+  config.are_outputs_on_device = false;
+  ICICLE_CHECK(keccakCUDA.hash(input.get(), chunk_size, config, output_main.get()));
+  END_TIMER(cuda_timer, "CUDA Keccak large time (on host memory)", true);
+  ASSERT_EQ(0, memcmp(output_main.get(), output_ref.get(), output_size * config.batch));
+
+  // test with device memory
+  std::byte *d_input = nullptr, *d_output = nullptr;
+  ICICLE_CHECK(icicle_malloc((void**)&d_input, total_size));
+  ICICLE_CHECK(icicle_malloc((void**)&d_output, output_size * config.batch));
+  ICICLE_CHECK(icicle_copy(d_input, input.get(), total_size));
+  config.are_inputs_on_device = true;
+  config.are_outputs_on_device = true;
+  START_TIMER(cuda_timer_device_mem);
+  ICICLE_CHECK(keccakCUDA.hash(d_input, chunk_size, config, d_output));
+  END_TIMER(cuda_timer_device_mem, "CUDA Keccak large time (on device memory)", true);
+  ICICLE_CHECK(icicle_copy(output_main_case_2.get(), d_output, output_size * config.batch));
+  ASSERT_EQ(0, memcmp(output_main_case_2.get(), output_ref.get(), output_size * config.batch));
 }
 
-// TODO: add tests for all hashes
+TEST_F(HashApiTest, sha3)
+{
+  auto config = default_hash_config();
```

Comment:
why are we testing it only on CPU?

Reply:
we test all devices including CUDA in the loop below